src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs

```diff
@@ -0,0 +1,300 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Executes analyzers on a document for computing local syntax/semantic diagnostics.
+    /// </summary>
+    internal sealed class DocumentAnalysisExecutor
+    {
+        private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>>? _lazySyntaxDiagnostics;
+        private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>>? _lazySemanticDiagnostics;
```

Comment:
Majority of the code in this file is existing code being moved here. The only new code is caching of per-analyzer syntax/semantic diagnostics in the executor, see methods `GetSyntaxDiagnosticsAsync` and `GetSemanticDiagnosticsAsync` below and the PERF notes in these methods.